components/admin_console/license_settings/modals/upload_license_modal.tsx

```diff
@@ -0,0 +1,281 @@
+// See LICENSE.txt for license information.
+
+import React, {useRef} from 'react';
+import {useSelector, useDispatch} from 'react-redux';
+
+import {FormattedDate, FormattedMessage} from 'react-intl';
+
+import {DispatchFunc} from 'mattermost-redux/types/actions';
+import {uploadLicense} from 'mattermost-redux/actions/admin';
+import {getLicenseConfig} from 'mattermost-redux/actions/general';
+import {getLicense} from 'mattermost-redux/selectors/entities/general';
+import {ClientLicense} from 'mattermost-redux/types/config';
+import {getCurrentLocale} from 'selectors/i18n';
+
+import {GlobalState} from 'types/store';
+
+import {isModalOpen} from 'selectors/views/modals';
+
+import GenericModal from 'components/generic_modal';
+import WomanArmOnTable from 'components/common/svg_images_components/woman_arm_on_table_svg';
+import HandsSvg from 'components/common/svg_images_components/hands_svg';
+import FileSvg from 'components/common/svg_images_components/file_svg';
+import LoadingWrapper from 'components/widgets/loading/loading_wrapper';
+
+import {ModalIdentifiers} from 'utils/constants';
+
+import {closeModal} from 'actions/views/modals';
+
+import {localizeMessage} from 'utils/utils';
+import {getMonthLong} from 'utils/i18n';
+
+import './upload_license_modal.scss';
+
+type Props = {
+    onExited?: () => void;
+}
+
+const UploadLicenseModal: React.FC<Props> = (props: Props): JSX.Element | null => {
+    const dispatch = useDispatch<DispatchFunc>();
+
+    const [fileObj, setFileObj] = React.useState<File | null>(null);
+    const [isUploading, setIsUploading] = React.useState(false);
+    const [serverError, setServerError] = React.useState<string | null>(null);
+    const fileInputRef = useRef<HTMLInputElement>(null);
+
+    const currentLicense: ClientLicense = useSelector(getLicense);
+    const locale = useSelector(getCurrentLocale);
+
+    const handleChange = () => {
+        const element = fileInputRef.current;
+        if (element === null || element.files === null || element.files.length === 0) {
+            return;
+        }
+        setFileObj(element.files[0]);
+        setServerError(null);
+    };
+
+    const handleSubmit = async (e: React.MouseEvent<HTMLButtonElement>) => {
+        e.preventDefault();
+        if (fileObj === null) {
+            return;
+        }
+
+        setIsUploading(true);
+        const {error} = await dispatch(uploadLicense(fileObj));
+
+        if (error) {
+            setFileObj(null);
+            setServerError(error.message);
+            setIsUploading(false);
+            return;
+        }
+
+        await dispatch(getLicenseConfig());
+        setFileObj(null);
+        setServerError(null);
+        setIsUploading(false);
+    };
+
+    const show = useSelector((state: GlobalState) => isModalOpen(state, ModalIdentifiers.UPLOAD_LICENSE));
+    if (!show) {
+        return null;
+    }
+
+    const handleOnClose = () => {
+        if (props.onExited) {
```

Comment:
Looked like one of the figma notes called for preventing closing the modal while upload is in progress. Could you implement that? I was able to simulate slow upload by using Chrome's performance tab in the inspector and creating a custom, really slow network profile.
![impossislow](https://user-images.githubusercontent.com/13738432/148122742-5f7504b4-bbed-4a78-a8c6-ed243a445f6a.png)

Reply:
sure!